plugin/backend_linux.go

```diff
@@ -797,6 +810,318 @@ func (pm *Manager) CreateFromContext(ctx context.Context, tarCtx io.ReadCloser,
 	return nil
 }
 
+// SavePlugin saves the plugin in outStream
+// The tar layout is as follows:
+// --- sha256ofconfig.json
+// --- sha256ofrootfs.tar.gz
+// --- manifest.json containing metadata of the form
+//	{
+//		"Name"  : pluginName
+//		"Config": sha256config.json
+//		"Layers": [
+//			"sha256rootfs.tar.gz"
+//		]
+//	}
+func (pm *Manager) SavePlugin(plugin string, outStream io.Writer) error {
+	srcP, err := pm.config.Store.GetV2Plugin(plugin)
+	if err != nil {
+		return err
+	}
+
+	pm.muGC.RLock()
+	defer pm.muGC.RUnlock()
+
+	ref, err := reference.ParseNormalizedNamed(plugin)
+	if err != nil {
+		return errors.Wrapf(err, "failed to parse reference %v", plugin)
+	}
+	if _, ok := ref.(reference.Canonical); ok {
+		return errors.Errorf("canonical references are not permitted")
+	}
+
+	// 1. create tempDir to save the tar
+	tempDir, err := ioutil.TempDir("", "docker-plugin-export-")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tempDir)
+
+	// 2. save config and rootfs into tmpdir
+	var manifest ocispecv1.Manifest
+
+	for i, dgst := range []digest.Digest{srcP.Config, srcP.Blobsums[0]} {
+		rdr, err := pm.blobStore.Get(dgst)
+		if err != nil {
+			return err
+		}
+		defer rdr.Close()
+
+		var mediaType string
+		if i == 0 {
+			mediaType = schema2.MediaTypePluginConfig
+			manifest.Config.Digest = dgst
+		} else {
+			mediaType = schema2.MediaTypeLayer
+		}
+
+		dest := filepath.Join(tempDir, "blobs", dgst.Algorithm().String(), dgst.Hex())
+		if err := os.MkdirAll(filepath.Dir(dest), 0755); err != nil {
+			return errors.Wrap(err, "error creating blob tree")
+		}
+
+		wtr, err := os.Create(dest)
+		if err != nil {
+			return err
+		}
+		defer wtr.Close()
+
+		size, err := io.Copy(wtr, rdr)
+		if err != nil {
+			return errors.Wrap(err, "error writing blob")
+		}
+
+		// set access and modified times of the manifest to be unix epoch, so that
+		// tar checksum is consistent.
+		if err := system.Chtimes(dest, time.Unix(0, 0), time.Unix(0, 0)); err != nil {
+			return errors.Wrap(err, "error resetting mod times for blob")
+		}
+
+		manifest.Layers = append(manifest.Layers, ocispecv1.Descriptor{
+			Digest:    dgst,
+			Size:      size,
+			MediaType: mediaType,
+		})
+
+		if mediaType == schema2.MediaTypePluginConfig {
+			manifest.Config.Size = size
+		}
+	}
+
+	manifest.Config.MediaType = mediaTypePluginConfig
+	manifest.Config.Annotations = map[string]string{"com.docker.plugin.ref.name": srcP.PluginObj.Name}
```

Comment:
Ah, yes. Of course. I was confused by the example in https://github.com/opencontainers/image-spec/blob/master/image-index.md#example-image-index